internal/dataplane/parser/translators/httproute_atc.go

```diff
@@ -0,0 +1,216 @@
+			path := ""
+			if match.Path != nil && match.Path.Value != nil {
+				path = *match.Path.Value
+			}
+			plugins := GeneratePluginsFromHTTPRouteFilters(filters, path, tags)
+			matchRoute.Plugins = plugins
+
+			routes = append(routes, matchRoute)
+		}
+		return routes, nil
+	}
+
+	// if we do not need to generate a kong route for each match, we OR matchers from all matches together.
+	routeMatcher := atc.And(atc.Or(generateMatchersFromHTTPRouteMatches(matches)...))
+	// add matcher from parent httproute (hostnames, protocols, SNIs) to be ANDed with the matcher from match.
+	matchersFromParent := matchersFromParentHTTPRoute(hostnames, ingressObjectInfo.Annotations)
+	for _, matcher := range matchersFromParent {
+		routeMatcher.And(matcher)
+	}
+
+	atc.ApplyExpression(&r.Route, routeMatcher, 1)
+	// generate plugins.
+	plugins := GeneratePluginsFromHTTPRouteFilters(filters, "", tags)
+	r.Plugins = plugins
+	return []kongstate.Route{r}, nil
+}
+
+func generateMatchersFromHTTPRouteMatches(matches []gatewayv1beta1.HTTPRouteMatch) []atc.Matcher {
+	ret := make([]atc.Matcher, 0, len(matches))
+	for _, match := range matches {
+		matcher := generateMatcherFromHTTPRouteMatch(match)
+		ret = append(ret, matcher)
+	}
+	return ret
+}
+
+func generateMatcherFromHTTPRouteMatch(match gatewayv1beta1.HTTPRouteMatch) atc.Matcher {
+	matcher := atc.And()
+
+	if match.Path != nil {
+		pathMatcher := pathMatcherFromHTTPPathMatch(match.Path)
+		matcher.And(pathMatcher)
+	}
+
+	if len(match.Headers) > 0 {
+		headerMatcher := headerMatcherFromHTTPHeaderMatches(match.Headers)
+		matcher.And(headerMatcher)
+	}
+
+	if match.Method != nil {
+		method := *match.Method
+		methodMatcher := methodMatcherFromMethods([]string{string(method)})
+		matcher.And(methodMatcher)
+	}
+	return matcher
+}
+
+func appendRegexBeginIfNotExist(regex string) string {
+	if !strings.HasPrefix(regex, "^") {
+		return "^" + regex
+	}
+	return regex
+}
+
+func pathMatcherFromHTTPPathMatch(pathMatch *gatewayv1beta1.HTTPPathMatch) atc.Matcher {
+	path := ""
+	if pathMatch.Value != nil {
+		path = *pathMatch.Value
+	}
+	switch *pathMatch.Type {
+	case gatewayv1beta1.PathMatchExact:
+		return atc.NewPredicateHTTPPath(atc.OpEqual, path)
+	case gatewayv1beta1.PathMatchPathPrefix:
+		if path == "" || path == "/" {
+			return atc.NewPredicateHTTPPath(atc.OpPrefixMatch, "/")
+		}
+		// if path ends with /, we should remove the trailing / because it should be ignored:
+		// https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.PathMatchType
+		path = strings.TrimSuffix(path, "/")
+		return atc.Or(
+			atc.NewPredicateHTTPPath(atc.OpEqual, path),
+			atc.NewPredicateHTTPPath(atc.OpPrefixMatch, path+"/"),
+		)
+	case gatewayv1beta1.PathMatchRegularExpression:
+		// TODO: for compatibility with kong traditional routes, here we append the ^ prefix to match the path from beginning.
+		// Could we allow the regex to match any part of path?
```

Comment:
I'd say that if the Gateway API does not strictly define what should be the behavior here, I'd stick to the one we already have for traditional routes. If we don't have a decision, maybe let's create an issue to track this and link to it herein the comment?

Reply:
https://github.com/Kong/kubernetes-ingress-controller/issues/3983